src/Tuura/Boolean.hs

```diff
@@ -11,76 +11,73 @@ import Data.Foldable
+genConcepts v o e
+    | length e == 1 = causes ++ " ~> " ++ effect
+    | otherwise       = "[" ++ causes ++ "]" ++ " ~|~> " ++ effect
   where
-    causes = unwords $ intersperse "," $ map (direction) e
+    causes = unwords $ intersperse "," $ map direction e
     direction (Literal a True)  = "rise " ++ a
     direction (Literal a False) = "fall " ++ a
-    op       = if (length e > 1) then " ~|~> " else " ~> "
-    effect   = if (v) then "rise out" else "fall out"
+    effect   = (if v then "rise " else "fall ") ++ o
 
-eval :: (Expr a) -> (a -> Bool) -> Bool
+eval :: Expr a -> (a -> Bool) -> Bool
 eval (Var a) f     = f a
 eval (Not a) f     = not (eval a f)
 eval (And a b) f   = eval a f && eval b f
 eval (Or a b) f    = eval a f || eval b f
 eval (SubExpr a) f = eval a f
 
-getValue :: Eq a => a -> [a] -> [Bool] -> Bool
-getValue var vars values = fromJust $ lookup var $ zip vars values
-
 simplifyDNF :: Ord a => DNF a -> DNF a
-simplifyDNF x = removeRedundancies $ removeSupersets x
+simplifyDNF x = DNF (removeRedundancies $ removeSupersets $ fromDNF x)
 
 simplifyCNF :: Eq a => CNF a -> CNF a
-simplifyCNF c = removeSupersets $ removeCancels vars c
+simplifyCNF c = CNF (removeSupersets $ fromCNF $ removeCancels vars c)
   where
-    vars = nub $ concatMap (map variable) c
+    vars = nub $ concatMap (map variable) (fromCNF c)
 
 removeCancels :: Eq a => [a] -> CNF a -> CNF a
 removeCancels [] whole = whole
 removeCancels (v:vs) whole = removeCancels vs newWhole
   where
     var = Literal v True
     nVar = Literal v False
-    relevant = filter (var `elem`) whole
+    relevant = filter (var `elem`) (fromCNF whole)
     nRelevant = map (replace [var] [nVar]) relevant
-    existInWhole = filter (`elem` whole) nRelevant
+    existInWhole = filter (`elem` fromCNF whole) nRelevant
     toBeRemoved = map (replace [nVar] [var]) existInWhole
     replacements = map (delete var) toBeRemoved
-    newWhole = whole ++ replacements
+    newWhole = CNF (fromCNF whole ++ replacements)
 
 convertCNFtoDNF :: Ord a => CNF a -> DNF a
-convertCNFtoDNF l = map (sort . nub) (sequence l)
+convertCNFtoDNF l = DNF $ map (sort . nub) (sequence (fromCNF l))
 
 -- Sort list of lists from largest length to shortest, then remove any lists
 -- that have shorter subsequences within the rest of the list.
-removeSupersets :: Eq a => CNF a -> CNF a
+removeSupersets :: Eq a => [[Literal a]] -> [[Literal a]]
 removeSupersets s = [ x | (x:xs) <- tails sortByLength, not (check x xs) ]
   where
     check current = any (`isSubsequenceOf` current)
     sortByLength  = sortBy (comparing $ negate . length) s
 
-removeRedundancies :: Eq a => CNF a -> CNF a
-removeRedundancies = filter (\ts -> all (\t -> not ((neg t) `elem` ts)) ts)
+removeRedundancies :: Eq a => [[Literal a]] -> [[Literal a]]
+removeRedundancies = filter (\ts -> all (\t -> neg t `notElem` ts) ts)
   where
     neg x = Literal { variable = variable x, polarity = not $ polarity x}
```

Comment:
Also, this looks like a generally useful function -- you need it above. Perhaps, call it `invert`?